### ISSUE TITLE
no-unused-vars: object key variables are considered unused

### ISSUE THREAD

**Issue by jnthnwn** [eslint/eslint#11083]
**What rule do you want to change?**
no-unused-vars should not consider variables used as object keys to be unused.

**Does this change cause the rule to produce more or fewer warnings?**
This will cause fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
I think this change should be added as the new default behaviour.

**Please provide some example code that this change will affect:**
```js
var A = 'a';

var obj = {
  A: 1,
};

function f(obj) {
  return obj.A;
}

console.log(f(obj));
```

**What does the rule currently do for this code?**
Currently, `no-unused-vars` reports that A is assigned a value and never used.

**What will the rule do after it's changed?**
The rule should not report an error for the above code snippet. It should not consider any variable used as a key in the object initializer syntax to be unused.

**Are you willing to submit a pull request to implement this change?**
Yes.

**Comment by platinumazure** [comment]
Hi @jnthnwn, thanks for the issue.

This is working as designed:

```js
var A = 'a'; // variable A

var obj = {
  A: 1, // object property A. Not the same as variable A
};

function f(obj) {
  return obj.A; // object property A, not variable A
}

console.log(f(obj));
```

The variable `A` is only assigned to, and it isn't used. So the rule is working correctly.

**Comment by jnthnwn** [comment]
Ah, that's embarrassing. Thanks for clarifying. I meant to use `{ [A]: 1 }`.
